cardano-crypto-class/src/Cardano/Crypto/Hash/Blake2b.hs

```diff
@@ -10,19 +10,41 @@ module Cardano.Crypto.Hash.Blake2b
   )
 where
 
+import Control.Monad (unless)
+import Cardano.Crypto.Libsodium.C
+
 import Cardano.Crypto.Hash.Class
-import qualified "cryptonite" Crypto.Hash as H
-import qualified Data.ByteArray as BA
+import Foreign.ForeignPtr (withForeignPtr, mallocForeignPtrBytes)
+import Foreign.Ptr (castPtr, nullPtr)
+import Foreign.C.Error (errnoToIOError, getErrno)
+import Data.Proxy
+import GHC.TypeLits (natVal)
+import GHC.IO.Exception (ioException)
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Data.ByteString as B
 
 data Blake2b_224
 data Blake2b_256
 
 instance HashAlgorithm Blake2b_224 where
   type SizeHash Blake2b_224 = 28
   hashAlgorithmName _ = "blake2b_224"
-  digest _ = BA.convert . H.hash @_ @H.Blake2b_224
+  digest _ = blake2b_libsodium (fromIntegral (natVal (Proxy::Proxy (SizeHash Blake2b_256))))
 
 instance HashAlgorithm Blake2b_256 where
   type SizeHash Blake2b_256 = 32
   hashAlgorithmName _ = "blake2b_256"
-  digest _ = BA.convert . H.hash @_ @H.Blake2b_256
+  digest _ = blake2b_libsodium (fromIntegral (natVal (Proxy::Proxy (SizeHash Blake2b_256))))
+
+blake2b_libsodium :: Int -> ByteString -> ByteString
+blake2b_libsodium size = \input -> unsafePerformIO $ do
+  output <- mallocForeignPtrBytes size
+  withForeignPtr output $ \output' -> do
+    B.useAsCStringLen input $ \(ptr, inputlen) -> do
+      res <- c_crypto_generichash_blake2b output' (fromIntegral size) (castPtr ptr) (fromIntegral inputlen) nullPtr 0 -- we used unkeyed hash
+      unless (res == 0) $ do
+        errno <- getErrno
+        ioException $ errnoToIOError "digest @Blake2b: crypto_generichash_blake2b" errno Nothing Nothing
+
+      B.packCStringLen (output', size)
```

Comment:
This is allocating a output buffer and then at the end coping out of that buffer to make a bytestring. We can do better and just allocate a single buffer which becomes the resulting bytestring. This saves an allocation and a copy.

It'll need some extra imports
```
import qualified Data.ByteString.Internal as B
```
and then, something like (untested)
```
blake2b_libsodium :: Int -> ByteString -> ByteString
blake2b_libsodium size input =
  B.unsafeCreate size $ \outptr ->
    B.useAsCStringLen input $ \(inptr, inputlen) -> do
      res <- c_crypto_generichash_blake2b
               (castPtr outptr) (fromIntegral size)
               (castPtr inptr) (fromIntegral inputlen)
               nullPtr 0 -- we used unkeyed hash
      unless (res == 0) $ do
        errno <- getErrno
        ioException $ errnoToIOError
                        "digest @Blake2b: crypto_generichash_blake2b"
                        errno Nothing Nothing
```